aws/credential_cache.go

```diff
@@ -97,10 +124,17 @@ func NewCredentialsCache(provider CredentialsProvider, optFns ...func(options *C
 // Returns and error if the provider's retrieve method returns an error.
 func (p *CredentialsCache) Retrieve(ctx context.Context) (Credentials, error) {
 	if creds, ok := p.getCreds(); ok && !creds.Expired() {
+		if p.options.EnableAsyncRefresh {
```

Comment:
good catch, indeed, although given that credential is the one responsible to define when it is expired, shall we consider that `ExpiryWindowJitterFrac` counts for this use case?
Do I understand correctly that this variable is here to "force" an earlier refresh of token while there is still one valid?

Reply:
> Do I understand correctly that this variable is here to "force" an earlier refresh of token while there is still one valid?

This variable being `ExpiryWindowJitterFrac`? The jitter just adds randomization to the expiry window.

> although given that credential is the one responsible to define when it is expired, shall we consider that `ExpiryWindowJitterFrac` counts for this use case?

I think the issue between this implementation and the token cache is the logic for determining when to refresh. The token cache doesn't adjust the expiration, it directly looks at the expiration window to determine when to attempt a refresh. This cache adjusts the expiration and lets that control when to refresh. The latter is simple but it doesn't lend itself to async refresh which really needs to be predicated on the expiry window (i.e. "I'm going to expire soon, I should try to refresh in the background before then so the next call for credentials succeeds and doesn't pay the latency cost"). For async refresh we need/want to know "am I in the expiry window?". 

It depends on what semantics we want to define here. I think simpler is better for understanding. I'd probably be fine with removing `AsyncRefreshMinimumDelay` and predicating simply off `ExpiryWindow` (adjusted for jitter). If we ask for credentials AND async refresh is enabled AND we are withing the expiry window then we try to refresh in the background. Otherwise it will be effectively a synchronous refresh (as it is now) if the credentials are expired, expiry window is 0, or async refresh is disabled. 

Thoughts?

Reply:
`If we ask for credentials AND async refresh is enabled AND we are withing the expiry window then we try to refresh in the background`: isn't it the behavior with this implementation when `AsyncRefreshMinimumDelay` is 0?

Reply:
Well right now the `ExpiryWindow` isn't taken into account at all because it's baked into the adjusted credentials expiration. `AsyncRefreshMinimumDelay` as it's currently implemented is basically a debounce for async refreshes. 

The additional `AsyncRefreshMinimumDelay` configuration option seems like an unnecessary control knob if the conditions described were met. It adds to the cognitive load in understanding this type for not much gained.

I think what we want to do here is enable async refresh based on just `ExpiryWindow` (w/jitter).

Here are the conditions to consider that I'm thinking of:

* `ExpiryWindow = 0`:  async refresh will have no effect and be same as today (synchronous)
* `ExpiryWindow > 0`: 
    * `EnableAsyncRefresh = false`: works as it does today, credentials expirations are adjusted and future retrieve calls will refresh the credentials before they expire but synchronous
    * `EnableAsyncRefresh = true`: try to async refresh in the background when we are within the expiry window such that the next time we ask for credentials we will have updated/refreshed creds with a further out expiration and not pay the latency cost inline with the operation being invoked. If the credentials are expired it would result in a synchronous refresh as it does today.

To enable this we need to track the original expiration of the credentials to know if we are within the expiry window (w/jitter). Otherwise the `ExpiryWindow` will be applied twice because the credentials are already adjusted to expire `Creds.Expiration = original expiration - (expiry window - jitter)`. More than likely this can just be stored as a tuple (creds, originalExpiration) in the atomic value that `creds` are stored in today but I haven't looked closely at it yet.

Reply:
The given definitions (1-3) are a confusing way to think about it IMO. I think I would phrase it with these definitions:

1. Expired - creds are considered expired (adjusted or otherwise)
2. Valid - not expired 
3. Within `ExpiryWindow` - creds are still valid but expiring soon
    * This is baked into `creds.Expired()` because the expiration times are adjusted by the `ExpiryWindow`. This is the "new" state to track to determine when to do an async refresh.

We shouldn't have to mess with the `creds.Expired()` logic, we just won't be able to rely on it for determining async refresh since the expiration time of the creds has already been adjusted by the `ExpiryWindow`. If this wasn't already the behavior and part of the public interface I would suggest changing this but we can't. 

> If a consumer wants to ensure to get systematically a valid token, the system will need to make sure that there is always a call to Retrieve during Pre-expiration window.

I'm not sure what you mean by this. Callers shouldn't care when they call `Retrieve`. If they want async refreshes to happen they should set a suitable `ExpiryWindow` and `EnableAsyncRefresh=true`. 


If this is still confusing let me know and I can try and write up a more detailed explanation of the changes to make.

Reply:
Updated with the logic described please review.
Not sure how consumer will react when getting a credential from cache which `creds.Expired()` returns true  :)